GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,466 @@
+        {
+            return this.localUpgradeServices.TrySetupToolsDirectory(out upgraderToolPath, out error);
+        }
+
+        private static bool TryGetPersonalAccessToken(string gitBinaryPath, string credentialUrl, ITracer tracer, out string token, out string error)
+        {
+            GitProcess gitProcess = new GitProcess(gitBinaryPath, null, null);
+            return gitProcess.TryGetCredentials(tracer, credentialUrl, out string username, out token, out error);
+        }
+
+        private string UnzipPackageToTempLocation()
+        {
+            string extractedPackagePath = this.localUpgradeServices.TempPath;
+            ZipFile.ExtractToDirectory(this.DownloadedPackagePath, extractedPackagePath);
+            return extractedPackagePath;
+        }
+
+        public class NugetUpgraderConfig
+        {
+            private readonly ITracer tracer;
+            private readonly LocalGVFSConfig localConfig;
+
+            public NugetUpgraderConfig(ITracer tracer, LocalGVFSConfig localGVFSConfig)
+            {
+                this.tracer = tracer;
+                this.localConfig = localGVFSConfig;
+            }
+
+            public NugetUpgraderConfig(
+                ITracer tracer,
+                LocalGVFSConfig localGVFSConfig,
+                string feedUrl,
+                string packageFeedName,
+                string feedUrlForCredentials)
+                : this(tracer, localGVFSConfig)
+            {
+                this.FeedUrl = feedUrl;
+                this.PackageFeedName = packageFeedName;
+                this.FeedUrlForCredentials = feedUrlForCredentials;
+            }
+
+            public string FeedUrl { get; private set; }
+            public string PackageFeedName { get; private set; }
+            public string FeedUrlForCredentials { get; private set; }
+
+            public bool TryLoad(out bool isEnabled, out bool isCorrectlyConfigured, out string error)
+            {
+                error = string.Empty;
+
+                string configValue;
+                string readError;
+
+                bool feedURLAvailable = false;
+                if (this.localConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeFeedUrl, out configValue, out readError))
+                {
+                    feedURLAvailable = !string.IsNullOrWhiteSpace(configValue);
+                }
+                else
+                {
+                    this.tracer.RelatedError(readError);
+                }
+
+                this.FeedUrl = configValue;
+
+                bool credentialURLAvailable = false;
+                if (this.localConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeFeedCredentialUrl, out configValue, out readError))
+                {
+                    credentialURLAvailable = !string.IsNullOrWhiteSpace(configValue);
+                }
+                else
+                {
+                    this.tracer.RelatedError(readError);
```

Comment:
Yeah... the code tries to read all config values before returning, so it can build a more accurate message. At the end of the method, we can see that this method only returns true if *all* config entries have values. It is a bit verbose - would a comment up front help?

Reply:
The NuGet config load code has been updated - hopefully it is now more clear...